### PR TITLE
feat: add a squash export

### PR DESCRIPTION
**Pull request by dktsni**
## Type of modification

* [ ] Breaking change
* [X] New Feature
* [ ] Bug Fix
* [ ] Chore (refactor, documentation, tests... all the changes with no impact on ARA functionalities.) 

## Changes description

This change will brings a new functionality exporter option to integrate functionalities in SquashTM.

## Technical description

* Update the list of available exporters to add a list of required fields which acts a list of required informations (by default this list is empty)
* Display this list dynamically in the export popup in the front end if it's not empty.

## PR CheckList

Please make sure your PullRequest respect all those items :

* [X] Your PR's title has the prefix : `feat:`, `fix:` or `chore:`
* [X] Unless your PR is related to a Hotfix (and approved by a ARA maintener), the targeted branch for it is the branch `version/X.Y.Z` and **not** `master`
* [X] You have asked a review from one of the ARA maintainer in your PR.
* [X] If your PR is related to an issue, add the issue's number in it.
* [x] All the code you added is documented.
* [ ] All the code you added is tested and the  tests are in success.
* [X] You already signed the [Contributor License Agreement](https://github.com/Decathlon/ara/blob/master/contributor-licence-agreement.adoc) and give us the document

### CODE REVIEW

File: client/src/components/functionality-export-popup.vue [hunk 1]
```diff
@@ -63,30 +88,65 @@ export default {
...
+      this.requiredFields = (currentExporter) ? currentExporter.requiredFields : undefined
+      this.selectedExporterId = id
+      this.fieldValue = {}
+      return id
+    },
+
+    decodeB64Stream (contentBase64) {
+      let binaryStr = window.atob(contentBase64)
+      let bStrLen = binaryStr.length
+      let contentArray = new Uint8Array(bStrLen)
```
Comment (omar-chahbouni-decathlon) [inline client/src/components/functionality-export-popup.vue]: In my opinion, the suffix `array` is not required: the variable name shouldn't care about what type (e.g. array, list, etc.) you use to store your data, rather `contents` (plural) could suffice.

File: database/instance/manage-db.sh [hunk 2]
```diff
@@ -80,8 +81,8 @@ function createContainer {
 	echo "[ARA] Creating the image..."
 	run 'docker build -t ara-db-image .'
 	echo "[ARA] Create the $CONTAINER_NAME container (data in $dataDir)"
-	docker_cmd=$(forgeCreationCommand)
-	docker_cmd="$docker_cmd -v /$dataDir:/var/lib/mysql" # Starting-slashes are for Windows compatibility
+	docker_cmd=$(forgeCreationCommand $dataDir)
```
Comment (omar-chahbouni-decathlon) [inline database/instance/manage-db.sh]: I think this fix another issue, maybe it should be in another PR ?

File: server/src/main/java/com/decathlon/ara/cartography/AraExporter.java [hunk 3]
```diff
@@ -36,7 +37,7 @@ public String getFormat() {
     }
 
     @Override
-    public byte[] generate(List<FunctionalityDTO> functionalities) {
+    public byte[] generate(List<FunctionalityDTO> functionalities, Map<String, String> requiredInfos) {
```
Comment (omar-chahbouni-decathlon) [inline server/src/main/java/com/decathlon/ara/cartography/AraExporter.java]: A documentation could help

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java [hunk 4]
```diff
@@ -1,16 +1,43 @@
...
  */
+@Slf4j
 public class SquashExporter extends Exporter {
+
+    private static final String PROJECT_NAME = "squash_project_name";
+    private static final String USER = "squash_user";
+    private static final Map<String,String> CRITICALITY_MAPPING = new HashMap<>();
+
+    static {
+        CRITICALITY_MAPPING.put("HIGH", "CRITICAL");
```
Comment (omar-chahbouni-decathlon) [inline server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: What do you think about using an Enum ? Final variables ?

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java [hunk 5]
```diff
@@ -27,8 +54,117 @@ public String getFormat() {
...
+    public byte[] generate(List<FunctionalityDTO> functionalities, Map<String, String> requiredInfos) {
+        String projectNameInSquash = requiredInfos.get(PROJECT_NAME);
+        String userWhichMakesTheImport = requiredInfos.get(USER);
+
+        HSSFWorkbook workbook = new HSSFWorkbook();
+        Sheet sheet = workbook.createSheet("REQUIREMENT");
+
+        this.createHeaderRow(sheet);
+        Date importDate = new Date();
+        int idx = 1;
```
Comment (omar-chahbouni-decathlon) [inline server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: Same comment as above

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java [hunk 6]
```diff
@@ -27,8 +54,117 @@ public String getFormat() {
...
+        this.addCellToRow(header, rowIdx++, "REQ_VERSION_STATUS");
+        this.addCellToRow(header, rowIdx++, "REQ_VERSION_DESCRIPTION");
+        this.addCellToRow(header, rowIdx++, "REQ_VERSION_#_TC");
+        this.addCellToRow(header, rowIdx++, "REQ_VERSION_#_ATTACHEMENT");
+        this.addCellToRow(header, rowIdx++, "REQ_VERSION_CREATED_ON");
+        this.addCellToRow(header, rowIdx++, "REQ_VERSION_CREATED_BY");
+        this.addCellToRow(header, rowIdx++, "REQ_VERSION_LAST_MODIFIED_ON");
+        this.addCellToRow(header, rowIdx++, "REQ_VERSION_LAST_MODIFIED_BY");
+        this.addCellToRow(header, rowIdx++, "REQ_VERSION_MILESTONE");
+        this.addCellToRow(header, rowIdx, "REQ_VERSION_CUF_<code du cuf>");
```
Comment (omar-chahbouni-decathlon) [inline server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: What about using an array storing all of your attribute (e.g. `"ACTION"`, `"PROJECT_ID"`, etc.), then looping through it meanwhile adding your cells ?

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java [hunk 7]
```diff
@@ -27,8 +54,117 @@ public String getFormat() {
...
+        this.addCellToRow(row, 4, "1");
+        this.addCellToRow(row, 5, "1");
+        this.addCellToRow(row, 7, criticity);
+        this.addCellToRow(row, 8, "CAT_UNDEFINED");
+        this.addCellToRow(row, 9, "WORK_IN_PROGRESS");
+        this.addCellToRow(row, 10, comment);
+        this.addCellToRow(row, 13, creationDate);
+        this.addCellToRow(row, 14, username);
+        this.addCellToRow(row, 15, creationDate);
+        this.addCellToRow(row, 16, username);
```
Comment (omar-chahbouni-decathlon) [inline server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: same comment as above, except maybe replacing the array by a map ?

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java [hunk 8]
```diff
@@ -27,8 +54,117 @@ public String getFormat() {
...
+        }
+        Long parentId = functionality.getParentId();
+        FunctionalityDTO parent = functionalities.stream()
+                .filter(f -> f.getId().equals(parentId))
+                .findFirst()
+                .orElse(null);
+        if (parent == null) {
+            return "";
+        }
+        return this.getFunctionalityPath(parent, functionalities) + "/" + parent.getName();
```
Comment (omar-chahbouni-decathlon) [inline server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: String.format ?

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java [hunk 9]
```diff
@@ -27,8 +54,117 @@ public String getFormat() {
...
+        Sheet sheet = workbook.createSheet("REQUIREMENT");
+
+        this.createHeaderRow(sheet);
+        Date importDate = new Date();
+        int idx = 1;
+        for (FunctionalityDTO functionality: functionalities) {
+            if (FunctionalityType.FUNCTIONALITY.name().equals(functionality.getType())) {
+                String criticity = CRITICALITY_MAPPING.get(functionality.getSeverity());
+                String funcName = functionality.getName();
+                String path = "/" + projectNameInSquash + this.getFunctionalityPath(functionality, functionalities);
```
Comment (omar-chahbouni-decathlon) [inline server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: String.format ?

File: client/src/components/functionality-export-popup.vue [hunk 10]
```diff
@@ -15,28 +15,53 @@
...
+        <div v-if="this.requiredFields && this.requiredFields !== {} ">
+          <Form>
+            <Form-item v-for="field in this.requiredFields" :key="field.id" :label="field.name + ' :'">
+              <div>
+                <form-field
+                  :field="field"
+                  ref="formField"
+                  v-model="fieldValue[field.id]" />
+              </div>
+              <div style="color: gray; line-height: 1.2; margin-top: 4px;">
```
[the remark this answers is not in the document]
Comment (dktsni) [reply client/src/components/functionality-export-popup.vue]: Here we are in a component, not in a page, so the CSS should be in the component IMHO to keep isolated.

File: client/src/components/functionality-export-popup.vue [hunk 11]
```diff
@@ -63,30 +88,65 @@ export default {
       }
       let idsToExport = []
       this.displayedFunctionalities.forEach(f => idsToExport.push(f.id))
+      let urlArgs = `exportType=${this.exporterId}&functionalities=${idsToExport.join(',')}`
+      if (this.fieldValue !== {}) {
+        let keys = Object.keys(this.fieldValue)
+        for (var idx in keys) {
```
[the remark this answers is not in the document]
Comment (dktsni) [reply client/src/components/functionality-export-popup.vue]: idx = index 

In Javascript a `for (var X in array)` loop isn't a foreach loop but an implicit `for(i=0; i < array.length; i++)` one :)

File: client/src/components/functionality-export-popup.vue [hunk 12]
```diff
@@ -63,30 +88,65 @@ export default {
...
+          let element = this.fieldValue[key]
+          urlArgs = `${urlArgs}&${key}=${element}`
+        }
+      }
       Vue.http
-        .get(api.paths.functionalities(this.projectCode) + `/export?exportType=${this.exporterId}&functionalities=${idsToExport.join(',')}`, api.REQUEST_OPTIONS)
+        .get(api.paths.functionalities(this.projectCode) + `/export?${urlArgs}`, api.REQUEST_OPTIONS)
         .then((response) => {
-          this.triggerDownload(response.bodyText)
+          console.log({ r: response })
```
[the remark this answers is not in the document]
Comment (dktsni) [reply client/src/components/functionality-export-popup.vue]: Done

File: client/src/components/functionality-export-popup.vue [hunk 13]
```diff
@@ -63,30 +88,65 @@ export default {
...
-    triggerDownload (content) {
-      const blob = new Blob([content], { type: 'text/plain' })
+    triggerDownload (response) {
+      let content = this.decodeB64Stream(response.bodyText)
+      const blob = new Blob([content], { type: 'application/octet-stream' })
       const event = document.createEvent('MouseEvents')
       let anchor = document.createElement('a')
-      anchor.download = `exportAraCartography-${this.$route.params.projectCode}.json`
+      let exporterFormat = ''
+      for (let idx in this.availableExporters) {
```
[the remark this answers is not in the document]
Comment (dktsni) [reply client/src/components/functionality-export-popup.vue]: idx = index

In Javascript a `for (var X in array)` loop isn't a foreach loop but an implicit `for(i=0; i < array.length; i++)` one :)

File: client/src/components/functionality-export-popup.vue [hunk 14]
```diff
@@ -63,30 +88,65 @@ export default {
...
+        if (this.availableExporters[idx].id === this.exporterId) {
```
[the remark this answers is not in the document]
Comment (dktsni) [reply client/src/components/functionality-export-popup.vue]: Respond is above

File: client/src/components/functionality-export-popup.vue [hunk 15]
```diff
@@ -63,30 +88,65 @@ export default {
...
       anchor.dispatchEvent(event)
     },
 
     cancelExport () {
       this.exporterId = ''
+      this.fieldValue = {}
+    },
+
+    changeCurrentExporter (id) {
+      let currentExporter = this.availableExporters.filter((ex) => ex.id === id)[0]
```
[the remark this answers is not in the document]
Comment (dktsni) [reply client/src/components/functionality-export-popup.vue]: Because if the array is empty, this method isn't called. It's the method used on the `change` event of the `select` tag alimented by this array and this `onchange` isn't called if the `select` is empty 😄

File: client/src/components/functionality-export-popup.vue [hunk 16]
```diff
@@ -63,30 +88,65 @@ export default {
...
+      this.requiredFields = (currentExporter) ? currentExporter.requiredFields : undefined
+      this.selectedExporterId = id
+      this.fieldValue = {}
+      return id
+    },
+
+    decodeB64Stream (contentBase64) {
+      let binaryStr = window.atob(contentBase64)
```
[the remark this answers is not in the document]
Comment (dktsni) [reply client/src/components/functionality-export-popup.vue]: In fact this variable contains a string not a stream. 

But I could change it if you really want to.

File: client/src/components/functionality-export-popup.vue [hunk 17]
```diff
@@ -63,30 +88,65 @@ export default {
...
+      let bStrLen = binaryStr.length
```
[the remark this answers is not in the document]
Comment (dktsni) [reply client/src/components/functionality-export-popup.vue]: In fact `binaryStr` contains a string not a stream.

But I could change it if you really want to.

File: client/src/components/functionality-export-popup.vue (hunk 1 above)
Comment (dktsni) [reply client/src/components/functionality-export-popup.vue]: I don't agree with you on this point. Here we manipulate a string by using directly the bytes in it.

When I use `contentArray` it's not for "an array of contents" but the "content in bytes array format", if you want I can change it to `contentByteArray`

File: database/instance/manage-db.sh (hunk 2 above)
Comment (dktsni) [reply database/instance/manage-db.sh]: with what purpose ?

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java (hunk 5 above)
Comment (dktsni) [reply server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: idx is the abbrev of index which is the array's index

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java (hunk 4 above)
Comment (dktsni) [reply server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: Those strings are used only here. I think that it will be over-engineered if we use final variables or enums here (+ the fact that the JVM will put those literal strings in its string pool).

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java (hunk 9 above)
Comment (dktsni) [reply server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: Despite the fact that String.format doesn't exists in Java (StringBuilder ...), we're effectively in a loop here so yep, totally agree with you 👍

Comment (dktsni) [reply server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: Done

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java (hunk 6 above)
Comment (dktsni) [reply server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: Ok for the headers :)

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java (hunk 7 above)
Comment (dktsni) [reply server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: For the moment a map will be less lisible and maintainable. If the format of the file change and more headers are added by Squash, why not 😄

File: server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java (hunk 8 above)
Comment (dktsni) [reply server/src/main/java/com/decathlon/ara/cartography/SquashExporter.java]: We're not in a loop here, so Java is able to detect the concatenation here and optimize it to a StringBuilder in the Bytecode.

And I found the String.format syntax less lisible than this (and again, String.format is absent of Java)

File: server/src/main/java/com/decathlon/ara/cartography/AraExporter.java (hunk 3 above)
Comment (dktsni) [reply server/src/main/java/com/decathlon/ara/cartography/AraExporter.java]: The documentation of the new param is made in the parent class (`Exporter`),  and with the `@Override` annotation, this method inherit the doc automagically 😄

**Comment by dktsni** [comment]
> Missing unit tests

Done now :) 
@omar-chahbouni-decathlon could you recheck now please ? 🙏

File: server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    private AraExporter sut;
+
+    @Test
+    public void getName_should_return_ARA() {
+        // Given
+        final String expected = "ARA";
+        // When
+        String actualName = this.sut.getName();
+        // Then
+        Assertions.assertThat(actualName).isEqualTo(expected);
```
Comment (omar-chahbouni-decathlon) [inline server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java]: Prefer static imports

File: server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+@RunWith(MockitoJUnitRunner.class)
+public class AraExporterTest {
+
+    @InjectMocks
+    private AraExporter sut;
+
+    @Test
+    public void getName_should_return_ARA() {
+        // Given
+        final String expected = "ARA";
```
Comment (omar-chahbouni-decathlon) [inline server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java]: good ;)

File: server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Test
+    public void getDescription_should_return_the_ara_description() {
+        // Given
+        final String expected = "Export this cartography to import it in another ARA project";
+        // When
+        String actualName = this.sut.getDescription();
+        // Then
+        Assertions.assertThat(actualName).isEqualTo(expected);
```
Comment (omar-chahbouni-decathlon) [inline server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java]: idem

File: server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Test
+    public void getFormat_should_return_json() {
+        // Given
+        final String expected = "json";
+        // When
+        String actualName = this.sut.getFormat();
+        // Then
+        Assertions.assertThat(actualName).isEqualTo(expected);
```
Comment (omar-chahbouni-decathlon) [inline server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java]: idem

File: server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+        final AraCartographyMapper mapper = new AraCartographyMapper();
+        final List<FunctionalityDTO> functionalities = new ArrayList<>();
+        for (int i=0; i < 10; i++) {
+            functionalities.add(this.create_dummy_functionality(i));
+        }
+        final byte[] expected = mapper.asString(functionalities).getBytes(StandardCharsets.UTF_8);
+        // When
+        byte[] actual = this.sut.generate(functionalities, null);
+        // Then
+        Assertions.assertThat(actual).containsExactly(expected);
```
Comment (omar-chahbouni-decathlon) [inline server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java]: idem

File: server/src/test/java/com/decathlon/ara/cartography/SquashExporterTest.java [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+        byte[] result = this.sut.generate(functionalities, requiredInfos);
+        // Then
+        Mockito.verify(this.sut, Mockito.times(1)).createHeaderRow(Mockito.any());
+        Mockito.verify(this.sut, Mockito.times(functionalities.size())).addRowToSheet(Mockito.any(), Mockito.anyInt(), Mockito.anyString(),
+                Mockito.anyString(), Mockito.anyString(), Mockito.anyString(), Mockito.any(), Mockito.anyString());
+        Assertions.assertThat(result).isNotEmpty();
+    }
+
+    @Test
+    public void listRequiredFields() {
```
Comment (omar-chahbouni-decathlon) [inline server/src/test/java/com/decathlon/ara/cartography/SquashExporterTest.java]: if you write your test methods names with snake_case, you should do the same here as well

File: server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java (hunk 18 above)
Comment (dktsni) [reply server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java]: I don't like static imports since they populate the PermGen space until the JVM exits. In this case it will populate the PermGen until our Maven process finishes.

By using standard imports, when the UT is finished, the import is unloaded :)

File: server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java (hunk 19 above)
Comment (dktsni) [reply server/src/test/java/com/decathlon/ara/cartography/AraExporterTest.java]: 👍

File: server/src/test/java/com/decathlon/ara/cartography/SquashExporterTest.java (hunk 23 above)
Comment (dktsni) [reply server/src/test/java/com/decathlon/ara/cartography/SquashExporterTest.java]: The lack of snake case here is because I only test what the documentation of the `listRequiredFields` method tells, the "passing case", there is only one path to test in this code, so I don't need to differentiate the test cases
